app/utils.ts

```diff
@@ -91,3 +89,69 @@ export const nameToParam = (name: string): string => {
 
 export const useIsDraggingComponent = () =>
     useSelector((state: EditorState) => state.isDragging)
+
+export const findMould = (state: EditorState, mouldName: string) =>
+    state.moulds.find((m) => m.name === mouldName)
+
+export const ensureMould = (state: EditorState, mouldName: string) => {
+    const mould = findMould(state, mouldName)
+    if (!mould) {
+        throw Error(`Mould '${mouldName}' was not found.`)
+    }
+
+    return mould
+}
+
+export const deleteMould = (state: EditorState, mouldName: string) => {
+    const mouldIndex = state.moulds.findIndex((m) => m.name === mouldName)
+    if (mouldIndex === -1) {
+        throw Error(`Mould '${mouldName}' was not found to delete.`)
+    }
+
+    return state.moulds.splice(mouldIndex, 1)
+}
+
+export const isDuplicateMould = (state: EditorState, mouldName: string) =>
+    !!state.moulds.find((m) => m.name === mouldName)
+
+export const getDefaultMouldName = (state: EditorState) => {
+    const len = state.moulds.length
+    let mouldName = `mould${len}`
+    if (isDuplicateMould(state, mouldName)) {
+        mouldName = `mould${len}` + nanoid(2)
+    }
+    if (isDuplicateMould(state, mouldName)) {
+        mouldName = `mould${len}` + nanoid(2)
+    }
+    if (isDuplicateMould(state, mouldName)) {
+        mouldName = `mould${len}` + nanoid(2)
+    }
+    // repeat only 2 times avoid endless loop
+    if (isDuplicateMould(state, mouldName)) {
+        throw Error(`Cannot find a available mould name.`)
+    }
```

Comment:
Repeating the check four times just reduces the probability of collision by four times, but it still exists and high.
As far as I can see, the `getDefaultMouldName` and `getDefaultStateName` functions are responsible for returning a unique name, but we don't really expect they throw an error

Why don't we just add something like auto incremented id, which will be more user friendly than unique generated id? For example, if you need to insert _mould4_ and it already exists, check wether exists _mould4_1_. If _mould4_1_ exists, check _mould4_2_ and so on

Or if we don't have time to do that, we could just increase the nanoid ID size each time, which will be reducing the collision probability significantly, for example 
```
while isDuplicateMould
  mouldName = `mould${len}_{nanoid(n++)}`

// with recusing the nanoid complexity 

while isDuplicateMould
  mouldName = `${mouldName}{nanoid(1)}`
```

P.S. A hash table could improve performance in both options so that we don't need to iterate the same array again and again

Reply:
Thank you. I think auto incremented id is good.

Reply:
Fixed and I'll merge it for others can deal with the changes soon.

If there are other issues, we can deal with them in a new pr